toolchain/parse/extract.cpp

```diff
@@ -0,0 +1,334 @@
+        if (it == end) {
+          *trace << "NodeIdNot " << T::Kind << " error: no more children\n";
+        } else {
+          *trace << "NodeIdNot error: unexpected " << T::Kind << "\n";
+        }
+      }
+      return std::nullopt;
+    }
+    if (trace) {
+      *trace << "NodeIdNot " << T::Kind << ": " << tree->node_kind(*it)
+             << " consumed\n";
+    }
+    return NodeIdNot<T>(*it++);
+  }
+};
+
+// Extract a `llvm::SmallVector<T>` by extracting `T`s until we can't.
+template <typename T>
+struct Extractable<llvm::SmallVector<T>> {
+  static auto Extract(const Tree* tree, Tree::SiblingIterator& it,
+                      Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<llvm::SmallVector<T>> {
+    if (trace) {
+      *trace << "Vector: begin\n";
+    }
+    llvm::SmallVector<T> result;
+    while (it != end) {
+      auto old_it = it;
+      auto item = Extractable<T>::Extract(tree, it, end, trace);
+      if (!item.has_value()) {
+        it = old_it;
+        break;
+      }
+      result.push_back(*item);
+    }
+    std::reverse(result.begin(), result.end());
+    if (trace) {
+      *trace << "Vector: end\n";
+    }
+    return result;
+  }
+};
+
+// Extract an `optional<T>` from a list of child nodes by attempting to extract
+// a `T`, and extracting nothing if that fails.
+template <typename T>
+struct Extractable<std::optional<T>> {
+  static auto Extract(const Tree* tree, Tree::SiblingIterator& it,
+                      Tree::SiblingIterator end, ErrorBuilder* trace)
+      -> std::optional<std::optional<T>> {
+    if (trace) {
+      *trace << "Optional: begin\n";
+    }
+    auto old_it = it;
+    std::optional<T> value = Extractable<T>::Extract(tree, it, end, trace);
+    if (value) {
+      if (trace) {
+        *trace << "Optional: found\n";
+      }
+      return value;
+    }
+    if (trace) {
+      *trace << "Optional: missing\n";
+    }
+    it = old_it;
+    return value;
+  }
+};
+
+// Extract a `tuple<T...>` from a list of child nodes by extracting each `T` in
+// reverse order.
+template <typename... T>
+struct Extractable<std::tuple<T...>> {
+  template <std::size_t... Index>
+  static auto ExtractImpl(const Tree* tree, Tree::SiblingIterator& it,
+                          Tree::SiblingIterator end, ErrorBuilder* trace,
+                          std::index_sequence<Index...>)
+      -> std::optional<std::tuple<T...>> {
+    std::tuple<std::optional<T>...> fields;
+    if (trace) {
+      *trace << "Tuple: begin\n";
+    }
+
+    // Use a fold over the `=` operator to parse fields from right to left.
+    [[maybe_unused]] int unused;
+    static_cast<void>(((std::get<Index>(fields) =
+                            Extractable<T>::Extract(tree, it, end, trace),
+                        unused) = ... = 0));
+
+    if (!(std::get<Index>(fields).has_value() && ...)) {
+      if (trace) {
+        *trace << "Tuple: error\n";
+      }
+      return std::nullopt;
+    }
```

Comment:
Along the failure path here, I think we need to reset `it` back to where it was before we started, otherwise (eg) an `optional<tuple>` or a `SmallVector<struct>` won't work properly.

Hm. Also, the pack fold tries to extract all elements even if an earlier extraction fails. That seems OK (though perhaps wasteful) when we're not tracing, but when we are tracing it will lead to additional trace output after the point of failure, and potentially a lot of it. Maybe we should stop extracting at the first failure:

```suggestion
    // Use a fold over the `=` operator to parse fields from right to left.
    [[maybe_unused]] int unused;
    bool ok = true;
    static_cast<void>(((ok &&
                        ok = (std::get<Index>(fields) =
                              Extractable<T>::Extract(tree, it, end, trace)).has_value(),
                        unused) = ... = 0));

    if (!ok) {
      if (trace) {
        *trace << "Tuple: error\n";
      }
      return std::nullopt;
    }
```